pkg/controller/function/function_controller.go

```diff
@@ -83,62 +99,171 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
-	fnConfigNamespaceEnv := os.Getenv("CONTROLLER_CONFIGMAP_NS")
+	// Get Function instance
+	fn := &runtimev1alpha1.Function{}
+	err := r.getFunctionInstance(request, fn)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
 
-	if len(fnConfigNameEnv) > 0 {
-		fnConfigName = fnConfigNameEnv
+		log.Error(err, "Error reading Function instance", "namespace", request.Namespace, "name", request.Name)
+		return reconcile.Result{}, err
 	}
 
-	if len(fnConfigNamespaceEnv) > 0 {
-		fnConfigNamespace = fnConfigNamespaceEnv
-	}
+	// Get Function Controller Configuration
 	fnConfig := &corev1.ConfigMap{}
-	err := r.Get(context.TODO(), types.NamespacedName{Name: fnConfigName, Namespace: fnConfigNamespace}, fnConfig)
-
-	if err != nil {
-		log.Error(err, "Unable to read Function controller config: %v from Namespace: %v", fnConfigName, fnConfigNamespace)
+	if err := r.getFunctionControllerConfiguration(fnConfig); err != nil {
 		return reconcile.Result{}, err
 	}
 
+	// Get a new *RuntimeInfo
 	rnInfo, err := runtimeUtil.New(fnConfig)
 	if err != nil {
-		fmt.Printf("Error in reading ConfigMap: %v", err)
+		log.Error(err, "Error while trying to get a new RuntimeInfo instance", "namespace", fnConfig.Namespace, "name", fnConfig.Name)
 		return reconcile.Result{}, err
 	}
-	// Fetch the Function instance
-	fn := &runtimev1alpha1.Function{}
-	err = r.Get(context.TODO(), request.NamespacedName, fn)
+
+	// Create Function's ConfigMap
+	foundCm := &corev1.ConfigMap{}
+	deployCm := &corev1.ConfigMap{}
+	_, err = r.createFunctionConfigMap(foundCm, deployCm, fn)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
 			return reconcile.Result{}, nil
 		}
-		// Error reading the object - requeue the request.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		log.Error(err, "function configmap can't be created. The function could have been deleted.", "namespace", deployCm.Namespace, "name", deployCm.Name)
+		return reconcile.Result{}, err
+	}
+
+	// Update Function's ConfigMap
+	if err := r.updateFunctionConfigMap(foundCm, deployCm); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		log.Error(err, "Error while trying to update Function's ConfigMap:", "namespace", deployCm.Namespace, "name", deployCm.Name)
+		return reconcile.Result{}, err
+	}
+
+	// Create function's image name
+	hash := sha256.New()
+	hash.Write([]byte(foundCm.Data["handler.js"] + foundCm.Data["package.json"]))
+	functionSha := fmt.Sprintf("%x", hash.Sum(nil))
```

Comment:
we could use this functionsha in the build-objects name as well...

Reply:
Now it is taking the functionSha and generating an short sha from the slice.